torch/_inductor/runtime/triton_helpers.py

```diff
@@ -384,3 +379,157 @@ def frexp(x):
+    )
+    left_idx = tl.reshape(left_idx, x.shape)
+    right_idx = tl.reshape(right_idx, x.shape)
+
+    # valid
+    if valid_mask is None:
+        left_valid_mask = tl.full(x.shape, True, tl.int1)
+        right_valid_mask = tl.full(x.shape, True, tl.int1)
+    else:
+        y_valid_mask = tl.reshape(valid_mask, shape)
+        left_valid_mask = tl.broadcast_to(
+            tl.sum(y_valid_mask * left_mask.to(tl.int8), 1)[:, None, :], shape
+        ).to(tl.int1)
+        right_valid_mask = tl.broadcast_to(
+            tl.sum(y_valid_mask * right_mask.to(tl.int8), 1)[:, None, :], shape
+        ).to(tl.int1)
+        left_valid_mask = tl.reshape(left_valid_mask, x.shape)
+        right_valid_mask = tl.reshape(right_valid_mask, x.shape)
+
+    # actual compare-and-swap
+    ix = x.to(idtype, bitcast=True)
+
+    if descending:
+        cond = left < right
+    else:
+        cond = left > right
+
+    if stable:
+        # When stable sorting, tie break by index
+        cond = cond | ((left == right) & (left_idx > right_idx))
+
+    cond = (right_valid_mask > left_valid_mask) | (
+        (right_valid_mask == left_valid_mask) & cond
+    )
+    cond = cond ^ flip
+    ret = ix ^ tl.where(cond, ileft ^ iright, tl.zeros_like(ix))
+    new_idxs = idxs ^ tl.where(cond, left_idx ^ right_idx, tl.zeros_like(idxs))
+    if valid_mask is None:
+        new_valid_mask = tl.full(x.shape, True, tl.int1)
+    else:
+        new_valid_mask = valid_mask ^ tl.where(
+            cond, left_valid_mask ^ right_valid_mask, tl.zeros_like(valid_mask)
+        )
+
+    return ret.to(x.dtype, bitcast=True), new_idxs, new_valid_mask
+
+
+@triton.jit
+def _bitonic_merge_with_index(
+    x,
+    idxs,
+    mask,
+    stage: tl.constexpr,
+    alternating: tl.constexpr,
+    n_dims: tl.constexpr,
+    stable: tl.constexpr,
+    descending: tl.constexpr,
+):
+    n_outer: tl.constexpr = x.numel >> n_dims
+    tl.static_assert(stage <= n_dims)
+    # flip denotes whether to re-arrange sub-sequences of elements in ascending or
+    # descending order.
+    # if flip = 00000000... then all elements will be re-arranged ascendingly at this stage
+    # if flip = 00110011... then all the elements will be re-arranged alternatingly (with
+    # a stride of 2) at this stage
+    if alternating:
+        shape: tl.constexpr = [n_outer * 2 ** (n_dims - 1 - stage), 2, 2**stage]
+        flip = tl.reshape(
+            tl.broadcast_to(tl.arange(0, 2)[None, :, None], shape), x.shape
+        )
+    else:
+        flip = False
+    # perform `stage` rounds of `compare-and-swap`
+    next_mask = mask
+    for i in tl.static_range(stage):
+        x, idxs, next_mask = _compare_and_swap_with_index(
+            x, idxs, mask, flip, i + (n_dims - stage), n_dims, stable, descending
+        )
+        if mask is not None:
+            mask = next_mask
+    return x, idxs, next_mask
+
+
+@triton.jit
+def sort_with_index(
```

Comment:
The write would be removed by inductor and triton can probably DCE the final steps of the index computation, but I do expect we could get a bigger win by having custom codegen for for sort without index. Though that wouldn't be able to support stable sorting since I'm using the indices to get a stable sort from the bitonic sort algorithm which is naturally unstable.

Reply:
Actually benchmarking it I see without indices is almost 2x faster so I guess triton is able to remove the index reductions entirely. That's pretty cool actually.